### PR TITLE
feat(table): add configurable fullSizeBp setting

### PR DESCRIPTION
**Pull request by pauleustice**
# Description

The LgTable currently always uses the non-mobile layout at and above the `md` breakpoint. However, if the table is placed in a container that does not fill the width of the screen then it can be desirable to show the mobile layout at larger screen widths.

This PR adds a new input which allows the consumer to specify at which breakpoint the full size layout should be displayed: 

`[fullSizeBp]="'sm'"` or `[fullSizeBp]="'md'"` or `[fullSizeBp]="'lg'"`

It defaults to using `md` if not provided, so should not introduce a breaking change. I also improved the LgTable storybook Knobs to make them easier to use.

## Requirements

Storybook link: https://deploy-preview-146--legal-and-general-canopy.netlify.app/

# Checklist:

- [x] The commit messages follow the [convention for this project](./blob/master/docs/CONTRIBUTING.md#conventional-commits)
- [x] I have provided an adequate amount of test coverage
- [x] I have provided a story in storybook to document the changes
- [x] I have provided documentation in the notes section of the story

### CODE REVIEW

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (pauleustice) [inline projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: I considered whether instances of these should be consolidated into a single scss file within `lib/table` but couldn't find a precedent for a similar approach. Please let me know if you think I should do this.

**Review by elenagarrone — DISMISSED** [review]
Nice one @pauleustice 👍  

For everyone else, I used `stack` as an input for the radio buttons component but Paul rightly pointed out that that's not really mobile friendly. I'll be updating the radio component in a separate PR so that it's consistent with this approach.

**Review by elenagarrone — DISMISSED** [review]

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (elenagarrone) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: I think it could probably go in an even more global place so that we can use the same breakpoints everywhere?

File: projects/canopy/src/lib/table/table-head-cell/table-head-cell.component.scss
Comment (aaronBery) [inline projects/canopy/src/lib/table/table-head-cell/table-head-cell.component.scss:3]: does this need to be re-declared throughout the tables sub components?

**Comment by aaronBery** [comment]
looks good and works well. just a few minor comments/questions @pauleustice  :-)

File: projects/canopy/src/lib/table/table-head-cell/table-head-cell.component.scss
Comment (pauleustice) [reply projects/canopy/src/lib/table/table-head-cell/table-head-cell.component.scss:3]: Agreed, it doesn't. Please see the comment I left: https://github.com/Legal-and-General/canopy/pull/146#discussion_r553318520

Just need to know where Canopy peeps would prefer I put it :)

File: projects/canopy/src/lib/table/table/table.component.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table/table.component.ts]: Resolved

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (pauleustice) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: @elenagarrone If there are other similar variables a global approach makes sense, otherwise I think within /lib/table would be better

Comment (elenagarrone) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: The radio also uses something similar`$stack-breakpoints: 'sm', 'md', 'lg' !default;`. 
Either you put `$full-size-breakpoints` in `/lib/table` and then once I modify the radio button I move it or you just move it now to a more global place and then I'll modify the radio variable name in a separate PR - together with some more bits to make it mobile friendly.

I don't mind which approach to be honest.

File: projects/canopy/src/lib/table/table.interface.ts
Comment (thatguynamedandy) [inline projects/canopy/src/lib/table/table.interface.ts:12]: We have an ongoing discussion around enums vs union types that we haven't really resolved yet. I wouldn't hold this PR up for it but any input you might have into it would be great.

File: projects/canopy/src/lib/table/table.notes.ts
Comment (thatguynamedandy) [inline projects/canopy/src/lib/table/table.notes.ts:79]: I'm not sure about the name of the property, I think it's the '**Bp**' that gets me as it might not be that clear what it stands for at first glance. I think this is only the second time we have used a property with a media breakpoint in it to define component behaviour (the radio stack being the other one), so we don't really have a convention yet.

Annoyingly I'm not sure of an alternative, from the design side the naming conventions are 'wide' and 'narrow', but I don't know if that's useful. 

`showWide="md"`
`fullSizeBreakpoint ="md"`
`wideAt="md"`
`wideBreakpoint="md"`

'wide' instead of `fullSize` is slightly more terse especially when used in the css, and it aligns with designs, but is it descriptive enough. Open to another opinion on that I think.

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (thatguynamedandy) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: Should they be the same? I'm just wondering if it's chance that on this occasion they happen to have the same properties.

> Nice one @pauleustice 👍
> 
> For everyone else, I used `stack` as an input for the radio buttons component but Paul rightly pointed out that that's not really mobile friendly. I'll be updating the radio component in a separate PR so that it's consistent with this approach.

@elenagarrone are you able to elaborate on that a bit? Do you mean you'll be changing the name of the input or is there more to it? I didn't quite pick up why it's not very mobile friendly.

**Review by thatguynamedandy — COMMENTED** [review]
Makes sense, I think it's a good addition, just a couple of minor comments before I thumb it :-)

**Comment by thatguynamedandy** [comment]
> Nice one @pauleustice 👍
> 
> For everyone else, I used `stack` as an input for the radio buttons component but Paul rightly pointed out that that's not really mobile friendly. I'll be updating the radio component in a separate PR so that it's consistent with this approach.

@elenagarrone are you able to elaborate on that a bit? Do you mean you'll be changing the name of the input or is there more to it? I didn't quite pick up why it's not very mobile friendly.

**Comment by pauleustice** [comment]
> > Nice one @pauleustice 👍
> > For everyone else, I used `stack` as an input for the radio buttons component but Paul rightly pointed out that that's not really mobile friendly. I'll be updating the radio component in a separate PR so that it's consistent with this approach.
> 
> @elenagarrone are you able to elaborate on that a bit? Do you mean you'll be changing the name of the input or is there more to it? I didn't quite pick up why it's not very mobile friendly.

@thatguynamedandy It's not mobile friendly as it is not applying the stacking (which should be the mobile/default layout of the component) by default. So the work would involve reversing the way the styles are applied, effectively.

File: projects/canopy/src/lib/table/table.interface.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table.interface.ts:12]: I think there are good reasons not to use enums (such as them being compiled), but I do like the syntax. Objectively, with all things considered, we should probably switch to union types IMO. For now though, this at least meets the precedent used in the radio group stack

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (pauleustice) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss:3]: @thatguynamedandy I think they should be kept separate, personally.

It does raise an interesting point that I considered when assessing the approach to take with this component. This method of using existing breakpoint names/sizes keeps things standardised (and in turn matches the grid layout and other media query-related styles), but on the other hand it doesn't allow the consumer to specify precise values based on their needs, and only supports one value.

My initial thinking before Elena told me about the existing radio stack approach was to make use of a `ResizeObserver` that listened to breakpoints specified within the input, but this was a much larger piece of work to consider. It would likely need a Canopy resize observer module that certain components (table and radio group at present) would depend on. Furthermore it isn't supported by IE.

It could have looked something like `<table lgTable [fullWidthAt]="{ sm: true, md: false }">` which would be nice. Maybe to consider in future?

File: projects/canopy/src/lib/table/table.notes.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table.notes.ts:79]: @thatguynamedandy I'm wary of 'wide' since it implies the width of the table rather than the layout used. But, I'm having a hard time coming up with better names. I prefer `fullSizeAt` or `fullSizeBreakpoint` but they still don't seem great.

File: projects/canopy/src/lib/table/table-cell/table-cell.component.scss
Comment (pauleustice) [reply projects/canopy/src/lib/table/table-cell/table-cell.component.scss]: Moved to `lib/table/table.breakpoints.scss`

File: projects/canopy/src/lib/table/table.notes.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table.notes.ts:79]: @thatguynamedandy @aaronBery How about `scaleUpAt="md"`?

File: projects/canopy/src/lib/table/table/table.component.ts
Comment (thatguynamedandy) [inline projects/canopy/src/lib/table/table/table.component.ts:92]: Sorry to be fussy but I wondered if this line should go in the `setter`. I found it odd that the `getter` returned the private variable but the `setter` didn't set it. Which does mean the ngOnInit would need to change as well 😬

I think that pattern that we normally use the one below in the alert component. I haven't considered it too much before but I guess there is probably some additional functional calls

https://github.com/Legal-and-General/canopy/blob/master/projects/canopy/src/lib/alert/alert.component.ts

File: projects/canopy/src/lib/table/table/table.component.ts
Comment (thatguynamedandy) [inline projects/canopy/src/lib/table/table/table.component.ts:39]: could the private variable shadow the name of the Input?
```suggestion
  private _showColumnsAt: TableColumnLayoutBreakpoints;
```

**Comment by aaronBery** [comment]
Sorry @pauleustice , just testing this again and I've found a slight inconsistency.

When in non-column mode on large mobile and the author label deselected the author label is not visible

Screenshot 2021-01-08 at 16 20 19

however when in column mode still on large mobile and the display author label still deselected the author label is now visible

Screenshot 2021-01-08 at 16 19 32

File: projects/canopy/src/lib/table/table/table.component.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table/table.component.ts]: Fixed

File: projects/canopy/src/lib/table/table/table.component.ts
Comment (pauleustice) [reply projects/canopy/src/lib/table/table/table.component.ts]: Fixed

**Review by thatguynamedandy — APPROVED** [review]

**Review by aaronBery — APPROVED** [review]
could potentially squash the commits and use the commit message from the first commit. functionality and code looks good :-)

**Pull request merged**
